src/aspire/source/coordinates.py

```diff
@@ -0,0 +1,485 @@
+    particles in a micrograph are represented by coordinates, but not yet
+    cropped out: `RelionCoordinateSource`. This class allows the output of
+    AutoPick and ManualPick jobs to be loaded into an ASPIRE source from a
+    single index STAR file (usually autopick.star)
+
+    Particle information is extracted from the micrographs and coordinate files
+    and put into a common data structure (self.particles)
+
+    The `_images()` method, called via `ImageSource.images()` crops
+    the particle images out of the micrograph and returns them as a stack.
+    This also allows the CoordinateSourceBase to be saved to an `.mrcs` stack.
+    """
+
+    def __init__(self, mrc_paths, coord_paths, particle_size, max_rows, dtype):
+        self.dtype = np.dtype(dtype)
+        self.particle_size = particle_size
+
+        # keep this list to identify micrograph paths by index rather than
+        # storing many copies of the same string
+        self.mrc_paths = mrc_paths
+
+        # The internal representation of micrographs and their picked coords
+        # is a list of tuples (index of micrograph, coordinate), where
+        # the coordinate is a list of the form:
+        # [lower left X, lower left Y, size X, size Y].
+        # the micrograph's filepath can be recovered from self.mrc_paths
+        self.particles = []
+        self.populate_particles(mrc_paths, coord_paths)
+
+        # get first micrograph and first coordinate to report some data
+        first_mrc_index, first_coord = self.particles[0]
+        first_mrc = self.mrc_paths[first_mrc_index]
+        with mrcfile.open(first_mrc) as mrc_file:
+            mrc_dtype = np.dtype(mrc_file.data.dtype)
+            shape = mrc_file.data.shape
+        if len(shape) != 2:
+            raise ValueError(
+                "Shape of mrc file is {shape} but expected shape of size 2. Are these unaligned micrographs?"
+            )
+        if self.dtype != mrc_dtype:
+            logger.warning(
+                f"dtype of micrograph is {mrc_dtype}. Will attempt to cast to {self.dtype}"
+            )
+
+        # save the shape to compare the rest of the mrcs against
+        self.mrc_shape = shape
+
+        # look at first coord to get particle size
+        # here we're checking the final coordinate of the first particle
+        # which is the Y-size of the box (the same as the X-size)
+        L = first_coord[3]
+
+        # total number of particles given in coord files
+        # before removing boundary particles and satisfying max_rows
+        original_n = len(self.particles)
+
+        logger.info(f"Micrograph size = {self.mrc_shape[1]}x{self.mrc_shape[0]}")
+        logger.info(f"Particle size = {L}x{L}")
+        self._original_resolution = L
+
+        # remove particles whose boxes do not fit at given particle_size
+        # and get number removed
+        boundary_removed = self.exclude_boundary_particles()
```

Comment:
My concern with this is the following: The specific particles in the source represented is essentially a function of `particle_size`. I think reporting the full size of the dataset given a particle size *before* `max_rows` is applied makes that super clear. Without the bigger picture someone might think that loading `max_rows=1000, particle_size=100` could be equivalent to `max_rows=1000, particle_size=300` are equivalent in some way, when its possible that that the latter would return particles from across say, 3 micrographs, rather than one.

Reply:
All of the coordinate files end up being opened anyway in `populate_particles`, because we don't yet know how many particles will be requested / across how many micrographs. This processing is happening with data already in memory.

Reply:
Okay. That's fair.  Before I consider resolving this, how long does it take to run this for 100k particles? 100? Maybe optimization isn't meritted.

Reply:
```
> from aspire.source import CentersCoordinateSource
> src = CentersCoordinateSource(files, particle_size=300)
> particles = src.particles
> len(particles)
194628
# I did this so that it would force some to be removed
> particles = [(idx, [coord[0],coord[1],600,600]) for idx, coord in particles]
> shape_y = src.mrc_shape[0]
> shape_y
3710
> shape_x = src.mrc_shape[1]
> shape_x
3838
> test_code = '''
out_of_range = []
for i, particle in enumerate(particles):
	start_x, start_y, size_x, size_y = particle[1] 
	if (start_x < 0 or start_y < 0 or (start_x + size_x >= shape_x) or (start_y + size_y >= shape_y)):
		out_of_range.append(i)
for j in reversed(out_of_range):
	particles.pop(j)
'''
> print(timeit.timeit(stmt=test_code, setup="from __main__ import particles, shape_x, shape_y", number=100))
4.780465215008007
> 4.780465215008007 / 100
0.047804652150080076






```

Reply:
Cool I guess I was confusing this step with reading from disk. Great, this is no problem. Thanks!